go/interpreter/lfvm/memory_test.go

```diff
@@ -47,3 +52,429 @@ func TestExpansionCosts(t *testing.T) {
+			if !errors.Is(err, test.expectedError) {
+				t.Errorf("unexpected error, want: %v, got: %v", test.expectedError, err)
+			}
+			if err == nil {
+				if !bytes.Equal(m.store[test.offset:test.offset+32], test.expectedData) {
+					t.Errorf("unexpected value, want: %v, got: %v", test.expectedData, m.store[test.offset])
+				}
+			}
+
+		})
+	}
+}
+
+func TestMemory_Set_SuccessfulCases(t *testing.T) {
+
+	tests := map[string]struct {
+		size   uint64
+		offset uint64
+	}{
+		"regular": {size: 8, offset: 0},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			m := NewMemory()
+			m.store = make([]byte, 8)
+			ctxt.gas = 100
+
+			data := []byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+			err := m.Set(test.offset, test.size, data)
+
+			if err != nil {
+				t.Fatalf("unexpected error, want: %v, got: %v", nil, err)
+			}
+
+			if m.Len() != test.size+test.offset {
+				t.Errorf("unexpected memory size, want: %d, got: %d", test.size+test.offset, m.Len())
+			}
+			if !bytes.Equal(m.store[test.offset:], data) {
+				t.Errorf("unexpected memory value, want: %x, got: %x", data, m.store[test.offset:])
+			}
+
+		})
+	}
+}
+
+func TestMemory_Set_ErrorCases(t *testing.T) {
+
+	tests := map[string]struct {
+		size     uint64
+		offset   uint64
+		expected error
+	}{
+		"size overflow": {
+			size:     math.MaxUint64,
+			offset:   1,
+			expected: errGasUintOverflow},
+		"offset overflow": {
+			size:     32,
+			offset:   math.MaxUint64,
+			expected: errGasUintOverflow},
+		"not enough memory": {
+			size:     32,
+			offset:   32,
+			expected: errSetMemTooSmall(8, 32, 32)},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			m := NewMemory()
+			m.store = make([]byte, 8)
+			ctxt.gas = 100
+
+			data := []byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+			err := m.Set(test.offset, test.size, data)
+
+			if !errors.Is(err, test.expected) {
+				t.Errorf("unexpected error, want: %v, got: %v", test.expected, err)
+			}
+
+		})
+	}
+}
+
+func TestMemory_CopyData(t *testing.T) {
+
+	baseStore := []byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+	baseTarget := []byte{0xfe, 0xdc, 0xba, 0x98, 0x76, 0x54, 0x32}
+
+	tests := map[string]struct {
+		target   []byte
+		store    []byte
+		offset   uint64
+		expected []byte
+	}{
+		"regular": {
+			target:   baseTarget,
+			store:    baseStore,
+			expected: baseStore[:len(baseTarget)]},
+		"offset bigger than memory": {
+			target:   baseTarget,
+			store:    baseStore,
+			offset:   9,
+			expected: bytes.Repeat([]byte{0}, len(baseTarget))},
+		"padding needed": {
+			target:   baseTarget,
+			store:    baseStore,
+			offset:   4,
+			expected: []byte{0x90, 0xab, 0xcd, 0xef, 0, 0, 0}},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			m := NewMemory()
+			m.store = test.store
+			target := test.target
```

Comment:
Since the `test.target` are reusing the same `baseTarget` slice, the effects of one test may influence other tests.